trillian/util/log_leaf.go

```diff
@@ -71,3 +50,57 @@ func ExtraDataForChain(cert ct.ASN1Cert, chain []ct.ASN1Cert, isPrecert bool) ([
 	}
 	return tls.Marshal(extra)
 }
+
+func BuildLogLeafWithHash(logPrefix string, merkleLeaf ct.MerkleTreeLeaf, leafIndex int64, cert ct.ASN1Cert, chain []ct.ASN1Cert, hash []byte, isPrecert bool) (trillian.LogLeaf, error) {
+	return buildLogLeaf(logPrefix, merkleLeaf, leafIndex, cert, chain, hash, isPrecert)
+}
+
+// ExtraDataForChainWithHash creates the extra data associated with a log entry as
+// described in RFC6962 section 4.6.
+func ExtraDataForChainWithHash(cert ct.ASN1Cert, chain []ct.ASN1Cert, hash []byte, isPrecert bool) ([]byte, error) {
+	var extra interface{}
+
+	if isPrecert {
+		// For a pre-cert, the extra data is a TLS-encoded PrecertChainEntry.
+		extra = ct.PrecertChainEntryHash{
+			PreCertificate:    cert,
+			IssuanceChainHash: hash,
+		}
+	} else {
+		// For a certificate, the extra data is a TLS-encoded:
+		//   ASN.1Cert certificate_chain<0..2^24-1>;
+		// containing the chain after the leaf.
+		extra = ct.CertificateChainHash{
+			IssuanceChainHash: hash,
+		}
+	}
+	return tls.Marshal(extra)
+}
+
+func buildLogLeaf(logPrefix string, merkleLeaf ct.MerkleTreeLeaf, leafIndex int64, cert ct.ASN1Cert, chain []ct.ASN1Cert, hash []byte, isPrecert bool) (trillian.LogLeaf, error) {
+	leafData, err := tls.Marshal(merkleLeaf)
+	if err != nil {
+		klog.Warningf("%s: Failed to serialize Merkle leaf: %v", logPrefix, err)
+		return trillian.LogLeaf{}, err
+	}
+
+	var extraData []byte
+	if hash == nil {
+		extraData, err = ExtraDataForChain(cert, chain, isPrecert)
+	} else {
+		extraData, err = ExtraDataForChainWithHash(cert, chain, hash, isPrecert)
+	}
+	if err != nil {
+		klog.Warningf("%s: Failed to serialize chain for ExtraData: %v", logPrefix, err)
+		return trillian.LogLeaf{}, err
+	}
+	// leafIDHash allows Trillian to detect duplicate entries, so this should be
+	// a hash over the cert data.
```

Comment:
This is true, but I think what you want to highlight something stronger: that it should be a hash over the cert data only, and *not* anything else, specifically not extraData.

Reply:
This is an existing code comment. We can create a separate pull request address this if the comment does not sound right.

https://github.com/google/certificate-transparency-go/blob/ea517d5f1795081b3d692db54beec25a507c62bf/trillian/util/log_leaf.go#L44-L46